numpy/core/src/multiarray/einsum_sumprod.c.src

```diff
@@ -10,38 +10,31 @@
 
 #define NPY_NO_DEPRECATED_API NPY_API_VERSION
 #define _MULTIARRAYMODULE
-
-#include <numpy/npy_common.h>
-#include <numpy/ndarraytypes.h>  /* for NPY_NTYPES */
 #include <numpy/halffloat.h>
-
 #include "einsum_sumprod.h"
 #include "einsum_debug.h"
+#include "simd/simd.h"
+#include "common.h"
 
-
-#ifdef NPY_HAVE_SSE_INTRINSICS
-#define EINSUM_USE_SSE1 1
+// ARM/Neon don't have instructions for aligned memory access
+#ifdef NPY_HAVE_NEON
+    #define EINSUM_IS_ALIGNED(x) 0
 #else
-#define EINSUM_USE_SSE1 0
-#endif
-
-#ifdef NPY_HAVE_SSE2_INTRINSICS
-#define EINSUM_USE_SSE2 1
-#else
-#define EINSUM_USE_SSE2 0
-#endif
-
-#if EINSUM_USE_SSE1
-#include <xmmintrin.h>
-#endif
-
-#if EINSUM_USE_SSE2
-#include <emmintrin.h>
-#endif
-
-#define EINSUM_IS_SSE_ALIGNED(x) ((((npy_intp)x)&0xf) == 0)
-
-/**********************************************/
+    #define EINSUM_IS_ALIGNED(x) npy_is_aligned(x, NPY_SIMD_WIDTH)
+#endif
+
+/**
+ * Unroll by four/eight scalars in case of:
+ *  - The main SIMD loop un-rolled by NPY_SIMD_WIDTH*(2|4), (2|4) represents
+ *    the number of times the loop that unrolled, Eg: for float32, 2 simd loop
+ *    is unrolled, for float64, 4 simd loop is unrolled, which may lead to
+ *    un-vectorize the remained scalars in bytes range <=NPY_SIMD_WIDTH*(2|4)
+ *    if The SIMD width is higher than 128bit, The performance loss on remained
+ *    arrays is nonnegligible, so we choose to use the compiler auto-vectorize.
+ *  - To give the chance to the compiler to
+ *    auto-vectorize in case of NPYV wasn't available.
```

Comment:
I think this reads a little better - let me know if I dropped some important information
```suggestion
 * This macro is used to enable a scalar loop which advances 4 elements at a
 * time, which appears after a main SIMD loop gated by `CHK` that unrolls by
 * `NPY_SIMD_WIDTH * unroll_by` elements, and before a non-unrolled scalar loop
 * that finishes up all the remaining scalars. The purpose of the unrolled loop
 * is to enable auto-vectorization in cases when all of the following are true:
 *
 *  - optimization is allowed
 *  - either:
 *    - we did not run the SIMD loop at all, due to NPV being disabled.
 *    - the SIMD loop was larger than 128bit, so there are likely to be many
 *      elements left to process.
```

Reply:
Thanks for the clear explanation.